code/lib/core-server/src/utils/StoryIndexGenerator.ts

```diff
@@ -270,6 +270,26 @@ export class StoryIndexGenerator {
     );
   }
 
+  /**
+   * Try to find the component path from a raw import string and return it in
+   * the same format as `importPath`.
+   *
+   * If no such file exists, assume that the import is from a package and
+   * return the raw path.
+   */
+  resolveComponentPath(rawComponentPath: Path, absolutePath: Path) {
+    const absoluteComponentPath = path.resolve(path.dirname(absolutePath), rawComponentPath);
+    const existing = ['', '.js', '.ts', '.jsx', '.tsx', '.mjs', '.mts']
+      .map((ext) => `${absoluteComponentPath}${ext}`)
+      .find((candidate) => fs.existsSync(candidate));
+    if (existing) {
+      const relativePath = path.relative(this.options.workingDir, existing);
+      return slash(normalizeStoryPath(relativePath));
+    }
+
+    return rawComponentPath;
+  }
```

Comment:
Would it make more sense to call `require.resolve(rawComponentPath, paths: [absolutePaths])`?

Reply:
I'll make a canary release and try this out in different scenarios. Can you list a few scenarios that this could have issues in @ndelangen?

Reply:
@yannbf let's say you have your stories all located in 1 directory.
I know... not the recommended approach.

Then you have your component files in some deep relative path on the fs, so you add a resolve to webpack so you can reference this location as:
```tsx
import { Button } from '@@components/button`;
```

That's an example of something that could break here.

Reply:
Thanks a lot Norbert! I tested the canary and here are my findings:

1. Maybe the docs page (if its meta has a component) should also have a componentPath, alongside importPath
<img width="416" alt="image" src="https://github.com/storybookjs/storybook/assets/1671563/71baaed9-3dcd-4395-8e13-423be6f4334a">

2. When you have aliased stories, e.g.
```ts
import { defineConfig } from "vite";
import react from "@vitejs/plugin-react";
import path from 'path';

export default defineConfig({
  plugins: [react()],
  resolve: {
    alias: {
      '@': path.resolve(__dirname, 'src'),
      '@stories': path.resolve(__dirname, 'src/stories'),
    },
  },
});
``` 

The value of `componentPath` ends up being the aliased one, like `@stories/Header`
<img width="416" alt="image" src="https://github.com/storybookjs/storybook/assets/1671563/420f3f07-39b8-40ad-a653-bcd4f452d091">

Maybe we just identify this as a limitation, else we'd have to do quite some extra work to figure out the aliasing, which can be complicated if people have merged configs from different files or a config file + (viteFinal|webpackFinal) combination